### PR TITLE
[WIP] Unwrap View in Reinterpret View

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89016

Previously, we would return False for `is_storage_and_layout` erroneously of Reinterpret Views when they wrapped a view, which would lead to an unneeded copy in requiring a specific layout. Since ReinterpretViews ignore the underlying object and just use the storage, we can unwrap view when creating it.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -1402,6 +1402,10 @@ class ReinterpretView(BaseView):
 
     layout: "Layout"
 
+    def __post_init__(self):
+        if isinstance(self.data, BaseView):
```
Comment (ngimel) [inline torch/_inductor/ir.py:1406]: how would this work? In the case that we've seen `View` converts 3d tensor to 2d, and then ReinterpretView permutes these 2 dimensions, relying on `.data` being 2d, but if instead `.data` is original 3d base, permuting 0th and 1st dim is something very different.

Comment (eellison) [reply torch/_inductor/ir.py:1406]: If you look at all the methods for `ReinterpretView`, `self.data` is just about ignored - we are only using the layout and the underlying storage. So, unwrapping the passed in view shouldn't affect correctness.

My understanding is that ReinterpretView would be initialized with a layout that already encodes that the tensor is 2d.

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -998,11 +998,11 @@ def as_storage_and_layout(x, freeze=True, want_contiguous=False, stride_order=No
                 x.data.decide_layout()
         return x, x.data.layout
     if isinstance(x, ReinterpretView):
+        # making the base of x contiguous or stride_ordered will not necessarily make
+        # the ReinterpretedView either, so dont pass along those arguments
```
Comment (ngimel) [inline torch/_inductor/ir.py:1002]: Ignoring those args seems fine for current usages, but it still feels fragile (nothing to be done in this PR however)

**Review by ngimel — APPROVED** [review]
Stamping if CI agrees, and regression is fixed.

**Pull request closed without merging**